### PR TITLE
fix level generation bug

### PR DESCRIPTION
**Pull request by friendlymatthew**
Screenshot 2024-05-21 at 2 01 35 PM

The `spawnLayer` was incorrectly computed. Additional asserts to check leveling.

### CODE REVIEW

File: pkg/hnsw/hnsw.go [hunk 1]
```diff
@@ -281,3 +281,35 @@ func (h *Hnsw) Insert(q Vector) error {
...
+	for k := range h.Nodes {
+		nodeIds = append(nodeIds, k)
+		num += 1
+	}
+
+	return nodeIds, num
+}
+
+// should never exceed MMAX, at level 0 shouldn't exceed MMAX0
+func (h *Hnsw) assert_num_neighbor_connections() (bool, error) {
```
Comment (kevmo314) [inline pkg/hnsw/hnsw.go:300]: Can you put this behind a compile guard to only run on debug builds?

File: pkg/hnsw/hnsw.go [hunk 2]
```diff
@@ -281,3 +281,35 @@ func (h *Hnsw) Insert(q Vector) error {
...
+func (h *Hnsw) numNodes() ([]NodeId, int) {
+	num := 0
+
+	var nodeIds []NodeId
+	for k := range h.Nodes {
+		nodeIds = append(nodeIds, k)
+		num += 1
+	}
+
+	return nodeIds, num
```
Comment (kevmo314) [inline pkg/hnsw/hnsw.go:296]:
```suggestion
	return nodeIds, len(nodeIds)
```

File: pkg/hnsw/hnsw.go [hunk 3]
```diff
@@ -281,3 +281,35 @@ func (h *Hnsw) Insert(q Vector) error {
 
 	return nil
 }
+
+// functions below are primarily used for testing
```
Comment (kevmo314) [inline pkg/hnsw/hnsw.go:285]: If they're only used for testing can they be in the `_test.go` file?

File: pkg/hnsw/hnsw_debug.go [hunk 4]
```diff
@@ -0,0 +1,28 @@
+//go:build debug
+// +build debug
```
Comment (friendlymatthew) [inline pkg/hnsw/hnsw_debug.go:2]: cmd: `go build -tags debug`

**Comment by friendlymatthew** [comment]
Closing due to #307

**Pull request closed without merging**